R/step-call-pivot_wider.R

```diff
@@ -0,0 +1,220 @@
+    new_vars <- mutate(data, .names_from = paste(!!!syms(names_from), sep = names_sep))
+    new_vars <- unique(pull(new_vars, .names_from))
+  } else {
+    new_vars <- unique(pull(data, !!sym(names_from)))
+  }
+
+  if (!is.null(names_glue)) {
+    glue_vars <- mutate(data, .names_from = glue(names_glue, .envir = .SD))
+    glue_vars <- unique(pull(glue_vars, .names_from))
+  }
+
+  if (length(values_from) > 1) {
+    new_vars <- lapply(values_from, function(.x) paste(.x, new_vars, sep = names_sep))
+    new_vars <- unlist(new_vars)
+  }
+
+  no_id <- length(id_cols) == 0
+
+  if (no_id) {
+    lhs <- "..."
+    new_vars <- c(".", new_vars)
+  } else {
+    lhs <- paste(id_cols, collapse = " + ")
+  }
+
+  vars <- c(id_cols, new_vars)
+
+  rhs <- paste(names_from, collapse = " + ")
+
+  dcast_form <- paste(lhs, rhs, sep = " ~ ")
+
+  args <- list(
+    formula = dcast_form,
+    value.var = values_from,
+    fun.aggregate = values_fn,
+    sep = names_sep,
+    fill = values_fill
+  )
+
+  # Clean up call args if defaults are used
+  args <- args[!vapply(args, is.null, lgl(1))]
+
+  if (names_sep == "_") {
+    args$sep <- NULL
+  }
+
+  out <- step_call(data, "dcast", args = args, vars = vars)
+
+  if (no_id) {
+    out <- select(out, -.)
+
+    new_vars <- new_vars[new_vars != "."]
+  }
+
+  if (!is.null(names_glue)) {
+    out <- step_call(
+      out,
+      "setnames",
+      args = list(old = new_vars, new = glue_vars),
+      vars = c(id_cols, glue_vars),
+      in_place = FALSE
+    )
+
+    # In case of names_sort = TRUE
+    new_vars <- glue_vars
+  }
+
+  if (nchar(names_prefix) > 0 && is.null(names_glue)) {
+    new_names <- paste0(names_prefix, new_vars)
+
+    out <- step_call(
+      out,
+      "setnames",
+      args = list(old = new_vars, new = new_names),
+      vars = c(id_cols, new_names),
+      in_place = FALSE
+    )
+
+    # In case of names_sort = TRUE
+    new_vars <- new_names
+  }
+
+  if (names_sort) {
+    cols_sorted <- c(id_cols, sort(new_vars))
+
+    out <- step_call(
+      out,
+      "setcolorder",
+      args = list(cols_sorted),
+      vars = cols_sorted,
+      in_place = FALSE
+    )
+  }
+
+  out <- step_repair(out, repair = names_repair)
+
+  out
+}
+
+globalVariables(c(".", ".names_from", "name", "value"))
+
+#' @export
+pivot_wider.data.table <- function(data,
+                                   id_cols = NULL,
+                                   names_from = name,
+                                   names_prefix = "",
+                                   names_sep = "_",
+                                   names_glue = NULL,
+                                   names_sort = FALSE,
+                                   names_repair = "check_unique",
+                                   values_from = value,
+                                   values_fill = NULL,
+                                   values_fn = NULL,
+                                   ...) {
+  data <- lazy_dt(data)
```

Comment:
Can you add one test for this, just to make sure it's correctly plumbed together? (it's easy to make mistakes in the tidy eval forwarding)